### ISSUE TITLE
Should it be an error to match a non-List to a List pattern?

### ISSUE THREAD

**Issue by leonsenft** [dart-lang/language#3049]
The same questions applies to all other similar patterns (maps, sets, records, etc).

The following snippet compiles without issue but will never match:

```dart
    if (node.argumentList case [_, final NamedExpression injector]) {
      buffer.write(injector.expression);
    }
```

That's because `node.argumentList` is not a `List`: it's an [ArgumentList](https://pub.dev/documentation/analyzer/latest/dart_ast_ast/ArgumentList-class.html) (which doesn't implement `List`).

What I _meant_ to write was:

```dart
    if (node.argumentList.arguments case [_, final NamedExpression injector]) {
      buffer.write(injector.expression);
    }
```

As I wanted to match against [arguments](https://pub.dev/documentation/analyzer/latest/dart_ast_ast/ArgumentList/arguments.html), which is a `List`.

Shouldn't the compiler/analyzer be able to trivially determine that the first example cannot possibly match?

Or is it allowed because we could have a secret type that implements `ArgumentList` **and** `List`–essentially the issue being that `ArgumentList` lacks the final class modifier?

**Comment by leonsenft** [comment]
> It requires a whole-program analysis to prove that there is not subclass of ArgumentList which implements List, and only our production compilers can ever assume that they know the whole program.

This is only true of non-sealed/final classes though, right? If the `ArgumentList` were final, wouldn't we be able to tell immediately that it's impossible for it to implement `List`?

**Comment by jakemac53** [comment]
> This is only true of non-sealed/final classes though, right? If the ArgumentList were final, wouldn't we be able to tell immediately that it's impossible for it to implement List?

It might be true of final but not sealed - you can extend the known subtypes of a sealed class.

**Comment by leonsenft** [comment]
> It might be true of final but not sealed - you can extend the known subtypes of a sealed class.

Outside of the defining library? I thought that was limited to the immediate library?

**Comment by lrhn** [comment]
If the class is `final` or private, and all subtypes of it are also `final` or private, then it's safe to assume that the class has no subtypes outside of the declaring library.

`List` is not such a type, but if `ArgumentList` is, then it should be possible to check that there is no subtype of `ArgumentList`, in the same library, which implements `List`.

And in that case, the analyzer is actually giving you a warning about the pattern not being able to match the value.
```dart
final class Floo {}
void main() {
  Floo floo = Floo() as dynamic;
  switch (floo) {
    case [int x]: print(x);
  }
}
```
gives me the warning
```
The matched value type 'Floo' can never match the required type 'List<Object?>'.
```

**Comment by leafpetersen** [comment]
> > It might be true of final but not sealed - you can extend the known subtypes of a sealed class.
> 
> Outside of the defining library? I thought that was limited to the immediate library?

To clarify, for both `sealed` and `final`:
  - Within the same library, a class marked `sealed` or `final` may be subclassed
    - If it is `final`, the subclass must be marked `base`, `sealed`, or `final`
  - The subclasses of a `sealed` or `final` class (inside of the same library by definition), may be extended outside of the library unless they are themselves marked `sealed` or `final`

**Comment by munificent** [comment]
See the related: https://github.com/dart-lang/language/issues/2632.

**Comment by Jetz72** [comment]
I think it'd be nice to have an analyzer setting to make it complain about this. If I'm testing whether something matches a particular list pattern, I'll more often than not be assuming it's a list and testing the content, rather than trying to simultaneously test the type and content.

One case where I'm particularly concerned is iterators. Seems like an easy rookie mistake to write something like:
```dart
if(strings.map((e) => e.toLowerCase()) case ["foo", "bar", "baz"])
```
It looks fine at a glance, but without a `toList` it'll always fail. At minimum I'd encourage a lint the specific case of static type `Iterable` against a List/Map/Set pattern.

**Comment by lrhn** [comment]
I don't think we should generally produce an error or warning when doing `case [...]` on something with static type `Iterable`.
Checking whether something is subtype of its static type is _the_ core use-case for type checks.

If the analyzer can recognize that the Iterable comes from one of the known lazy functions on Iterable, _then_ it can heuristically _assume_ that it won't be a list. That sounds reasonable. A kind if "precise type" which cannot be inferred, or ensured, from the types alone.

**Comment by Jetz72** [comment]
> Checking whether something is subtype of its static type is _the_ core use-case for type checks.

True, but that goes back to the question: does the user actually intend or expect the case test to act as a type check? When you write `if(strings[0] == "foo" && strings[1] == "bar")`, the intention is straightforward. If the analyzer isn't sure that `strings` has an `[]` operator, then it'll let you know that, and you have to actually consider what happens in the event that you're working with something other than a list. Maybe you just add in the explicit `strings is List` check at the start of the if statement, or maybe you add a separate branch to handle that case with its own logic. Whatever is done, the edge case of "`strings` is not a List" gets the consideration its due.

But if someone's written `if(strings case ["foo", "bar", _])` as a shorthand for testing their items in what they assume will be a list, it may never cross their mind that `strings` could be something else. If that's not what they intend, it could lead to some rather insidious bugs that either take a while to puzzle out, or slip through to production code and go unnoticed for a very long time. 

I'm concerned that this will become a common mistake for people to make when they're not experienced with this feature.

**Comment by lrhn** [comment]
> "does the user actually intend or expect the case test to act as a type check?"

If they write `case`, we assume that they do want to check the type, because there is no way to see otherwise. It's precisely what they would write *if* they want a type-check. It's the official way to ask for a type check.

The example here is precisely the kind of use-case you'd see:
```dart
String abbreviateCollection(Iterable collection) =>
  switch (collect) {
   [] => "[]",
   [var only] => "[$only]",
   [var first, ..., var last] => "[$first, ..., $last]",
   Set set when set.isEmpty => "{}",
   Set set when set.length == 1 => "{${set.first}}",
   Set set => "{${set.first}, ...}",
   _ => "(...)"
};
```

The type alone doesn't say anything.

The extra knowledge that `foo.map(...)` is almost always a non-`List` iterable might be useful, and the analyzer would be in a position to give a warning, but it cannot be part of the language.
